Enhancement/language selector

ArchitectedDev opened the pull request:
This seems like a good starting point for the feature, although there's likely optimisations that can be made.

There are a few other pieces that I want to continue to work on, so we can hold off on the merge if desired, but any thoughts on improvements are welcome.

- Identify and translate remaining hardcoded text in templates.
- Store and retrieve selected language in cookie (or persist to profiled or anonymous user entity in OrderCloud).
- Update docs where relevant.

ArchitectedDev committed:
- Initial implementation for multi-lingual support for static content in buyer and seller apps.
- Moved and renamed translation files into API deployment folder.
- Cleaned up the seed process to automatically deploy any and all translations files in the API deployment folder.
- Removed StorageAccountSeedSettings from the EnvironmentSeedRequest model as it duplicates StorageAccountSettings.cs
- Created devops/translations endpoint to allow translation files to be manually uploaded.
- Added hard-coded language selector controls to the buyer and seller apps to be replaced with dynamically populated selections.
- Added supportedLanguages and defaultLanguage properties to buyer and seller apps.

ArchitectedDev committed:
Updated language selector controls to be populated dynamically from configured locales.

crhistianramirez commented on src/Middleware/src/Headstart.API/Controllers/DevOpsController.cs:
```diff
@@ -0,0 +1,28 @@
+        public async Task UploadTranslations([FromForm] UploadTranslationsRequest translationsRequest)
+        {
+            await command.UploadTranslationsFile(translationsRequest);
+        }
```
I think this method is good and will be a lot more useful if we were to associate some kind of admin UI with the functionality here but we probably also want another command and endpoint that updates all translation files based on what is in the current source code. Two reasons for this

1. Anyone that is following along with headstart updates can call this endpoint to quickly retrieve updated translations
2. There is a use case I see quite commonly where translations are tied to a release and updating the translations  is generally a developer task. Such a flow could be created by calling this endpoint as part of the final step in a release pipeline.

ArchitectedDev replied:
Good points here.

1. Adding a 'global' configuration section in the admin app will allow admin business users to perform actions that currently require developer intervention.
2. I briefly considered this, but didn't review the release pipeline to understand actions currently being performed. In any case, this is easy additional endpoint is easy to add.

ArchitectedDev replied:
The release pipeline will need to be manually updated to resolve [Update translation files on deploy based on the saved translation file](#294). Additional details in linked issue.

crhistianramirez commented on src/UI/Buyer/src/app/app.module.ts:
```diff
@@ -433,8 +433,9 @@ export class AppModule {
-    translate.setDefaultLang('en')
-    translate.use('en')
+    translate.addLangs(this.appConfig.supportedLanguages)
+    translate.setDefaultLang(this.appConfig.defaultLanguage)
```
Looking at the library we're using we might be able to make this a little bit smarter by detecting the user's language from the browser (`getBrowserCultureLang` and `getBrowserLang`) and using that, and if the language isn't supported then falling back to the default language. That might look something like this (untested):

```typescript
translate.addLangs(this.appConfig.supportedLanguages)
const browserCultureLang = translate.getBrowserCultureLang();
const browserLang = translate.getBrowserLang();
if (this.appConfig.supportedLanguages.includes(browserCultureLang) {
   translate.setDefaultLang(browserCultureLang);
} else if (this.appConfig.supportedLanguages.includes(browserLang ) {
   translate.setDefaultLang(browserLang );
} else {
   translate.setDefaultLang(this.appConfig.defaultLanguage)
}

```

crhistianramirez replied:
~It probably also makes sense to persist the language choice in both the cookies and on the user object. We need it on cookies because user might not be logged in, if they are though prefer to retrieve from user object and then finally if neither then use the above logic to persist for the first time~

^ I see in your PR summary you've already considered this

ArchitectedDev replied:
Sounds good, I will look at adding this improvement.

ArchitectedDev replied:
This has been completed and has been commented on separately, so I will resolve this.

crhistianramirez commented on src/UI/Buyer/src/app/components/layout/app-header/app-header.component.html:
```diff
@@ -41,6 +41,31 @@
       </div>
 
+      <!-- Language Selector -->
+      <ul class="navbar-nav" *ngIf="screenSize !== 'xs' && screenSize !== 'sm'">
```
We'll want to account for mobile as well before calling this complete

ArchitectedDev replied:
I agree, although I was just following the current pattern with the other controls, which doesn't appear to support mobile at this time, so I will have to identify the UX for this as well as the account controls.

crhistianramirez commented on src/UI/Buyer/src/app/config/app.config.ts:
```diff
@@ -13,6 +13,8 @@ export const ocAppConfig: AppConfig = {
   marketplaceName: environment.marketplaceName,
   translateBlobUrl: environment.translateBlobUrl,
+  supportedLanguages: environment.supportedLanguages,
+  defaultLanguage: environment.defaultLanguage,
```
We should add these variables to docker as well, or at the very least add the languages that are supported out of the box in Buyer/src/assets/appConfigs and Seller/src/assets/appConfigs and english as a default

ArchitectedDev replied:
I had performed a quick search for docker variables, but didn't see any, but upon further review I see that only a subset of properties are included in docker. I assume we should consider updating the docker variables to be 1:1 with the buyer and seller configs?

For the buyer apps, I think it would be more optimised by making it configurable in the seller app against the API Client xp as it's already representing the storefront with the IsStorefront xp.

crhistianramirez replied:
> For the buyer apps, I think it would be more optimised by making it configurable in the seller app against the API Client xp

I'm not opposed to adding this as an enhancement and would let us do some cool things like changing themes on the fly but we'll need to consider how to map a buyer application to a specific API client. The easiest would be to have only the clientID stored as a frontend configuration. I've also thought about something that is more dynamic like mapping a domain to an API client. That requires a bit more set up the first time around but has some powerful implications. For example if someone has a wildcard domain they could theoretically pop up new buyers pretty much at will without any kind of deployment process.

crhistianramirez commented:
Very good start indeed. I'll leave it up because I think it would be nice to package everything in a single clean PR. Also, I understand this is a WIP, my comments were really just identifying the things we'll want to address so I don't forget more than anything. I'd like to pivot to this as soon as I finish bundles (hopefully today) so we can tag team some of the things I brought up.

ArchitectedDev commented:
Great feedback @crhistianramirez. Agree to merge only once everything is complete. I am not sure if there is a better way to collaborate on WIPs rather than a PR, so let me know if you are aware of an alternate. Thanks.

crhistianramirez commented:
> I am not sure if there is a better way to collaborate on WIPs rather than a PR

I like the PR as a tool for collaboration as long as you don't mind me committing to the same branch. We'll just need to commit small and often and can update this thread when either of us needs one of the other to look at something.

ArchitectedDev committed:
- Updated the translations endpoints to support individual file upload and 'sync' (push from deployed i18n folder) operations.
- Extracted FileUpload class out of ShipmentController for general consumption.

ArchitectedDev committed:
Added support for storing and retrieving the selected language from a cookie.

ArchitectedDev committed:
added language configs to buyer and sell apps' configs

ArchitectedDev committed:
- Added support for storing/retrieving selected language to/from profiled user.
- Slightly improved approach to handling language selection.

crhistianramirez commented on src/UI/Buyer/src/app/app.module.ts:
```diff
@@ -540,6 +544,30 @@ export class AppModule {
+    } else if (languages.includes(this.appConfig.defaultLanguage)) {
+      this.translate.setDefaultLang(this.appConfig.defaultLanguage)
+    } else if (languages.length > 0) {
+      this.translate.setDefaultLang(languages[0])
```
Can we throw an error at the top of this function if there are no languages?

ArchitectedDev replied:
Sure, I will add defensive programming here. I wasn't sure if we were going to fail fast or just silently handle this by letting the placeholder references display. Are you aware of a preferred approach or whether their are any post-deployments checks to identify the app is running successfully?

ArchitectedDev replied:
Errors are now thrown when `supportedLanguages` or `defaultLanguage` are not defined in the config, or when no language can be identified for use (which should never occur).

The initial errors do prevent the sites from loading, so we may want to handle showstopping errors by rendering some kind of technical difficulties error page, however I am not sure the best way to manage this at this time.

crhistianramirez replied:
A dedicated error page could be nice, but maybe just a toaster that needs to be manually dismissed would suffice for now

crhistianramirez commented on src/UI/Buyer/src/app/services/auth/auth.service.ts:
```diff
@@ -134,6 +140,16 @@ export class AuthService {
+      this.translate.use(user.xp?.Language);
+    } else {
+      this.cookieService.remove(this.languageCookieName)
+      this.translate.use(this.translate.defaultLang);
```
Shouldn't the logic here be similar to where we try to use the user's location to retrieve the most relevant language?

ArchitectedDev replied:
Agreed, I have been working creating a shared service specifically to manage getting and setting the language to adhere to the single responsibility principle.

ArchitectedDev replied:
I have moved this implementation into language-selector.service.ts. One for the buyer app and one for the seller app as each app has their own specific services, helpers, etc. that are not reusable.

ArchitectedDev committed:
Further improvements to making the language selection (storage/retrieval) an external service.

ArchitectedDev committed:
Updated the language selector service to support anonymous users and log out events.

ArchitectedDev committed:
Updated the seller app with its own language selector service.

ArchitectedDev committed:
- Resolved issues seller language selector implementation.
- Updated user patch to only include the language xp.

ArchitectedDev committed:
Updated the buyer and seller app READMEs with latest configuration details.

ArchitectedDev committed:
- Updating documentation to cover how to import the middleware API in postman.
- Corrected a few typos.

ArchitectedDev committed:
Replacing static text with translated text.

ArchitectedDev committed:
Added mobile navbar support to buyer app.

ArchitectedDev committed:
- Updated functionality to collapse mobile navbar menu on interaction with navbar items.
- Updated language selector service to utilise current user service to be closer to have the current user updated when language is patched to profile.

ArchitectedDev committed:
Added language properties to docker files.

ArchitectedDev committed:
Updated documented endpoint for updating translation files.
